applications/FluidDynamicsApplication/custom_utilities/cfd_utilities.cpp

```diff
@@ -0,0 +1,386 @@
+        dx = value - y_plus;
+
+        if (std::abs(dx) < Tolerance)
+            return y_plus;
+
+        y_plus = value;
+    }
+
+    KRATOS_WARNING("CalculateLinearLogarithmicWallFunctionBasedYPlusLimit")
+        << "Logarithmic y_plus limit reached max iterations with dx > "
+           "Tolerance [ "
+        << dx << " > " << Tolerance << ", MaxIterations = " << MaxIterations << " ].\n";
+    return y_plus;
+}
+
+double CalculateLinearLogarithmicWallFunctionBasedYPlusAndUtau(
+    array_1d<double, 3>& rFrictionVelocity,
+    const array_1d<double, 3>& rWallVelocity,
+    const array_1d<double, 3>& rNormal,
+    const double KinematicViscosity,
+    const double WallHeight,
+    const double VonKarman,
+    const double WallSmoothness,
+    const int MaxIterations,
+    const double Tolerance)
+{
+    KRATOS_TRY
+
+    // calculating unit normal since rNormal contains surface area of the condition
+    const double surface_area = norm_2(rNormal);
+    const array_1d<double, 3>& unit_normal = rNormal / surface_area;
+
+    // calculate tangential velocity
+    noalias(rFrictionVelocity) =
+        rWallVelocity - unit_normal * inner_prod(rWallVelocity, unit_normal);
+    const double wall_velocity = norm_2(rFrictionVelocity);
+
+    // calculate linear log region limit
+    const double limit_y_plus = CalculateLinearLogarithmicWallFunctionBasedYPlusLimit(
+        VonKarman, WallSmoothness, MaxIterations, Tolerance);
+
+    // linear region
+    double u_tau = std::sqrt(wall_velocity * KinematicViscosity / WallHeight);
+    double y_plus = u_tau * WallHeight / KinematicViscosity;
+    const double inv_kappa = 1.0 / VonKarman;
+
+    // log region
+    if (y_plus > limit_y_plus)
+    {
+        int iter = 0;
+        double dx = 1e10;
+        double u_plus = inv_kappa * std::log(y_plus) + WallSmoothness;
+
+        while (iter < MaxIterations && std::abs(dx) > Tolerance * u_tau)
+        {
+            // Newton-Raphson iteration
+            double f = u_tau * u_plus - wall_velocity;
+            double df = u_plus + inv_kappa;
+            dx = f / df;
+
+            // Update variables
+            u_tau -= dx;
+            y_plus = WallHeight * u_tau / KinematicViscosity;
+            u_plus = inv_kappa * std::log(y_plus) + WallSmoothness;
+            ++iter;
+        }
+
+        KRATOS_WARNING_IF("CFDUtilities", iter >= MaxIterations)
+            << "CalculateLinearLogarithmicWallFunctionBasedYPlusAndUtau "
+               "couldn't converge Newton-Raphson. Residual is "
+            << dx << std::endl;
+    }
+
+    noalias(rFrictionVelocity) =
+        rFrictionVelocity *
+        (-1.0 * u_tau /
+         (wall_velocity <= std::numeric_limits<double>::epsilon() ? 1.0 : wall_velocity));
+
+    return y_plus;
+
+    KRATOS_CATCH("");
+}
+
+void CalculateYPlusAndUTauForConditions(
+    ModelPart& rModelPart,
+    const Variable<double>& rKinematicViscosityVariable,
```

Comment:
Note what we discussed about Kinematic vs. Dynamic viscosity. Some years ago we decided to favor the use of the dynamic viscosity to be consistent among the different elements. I suggest following this in here.